### DISCUSSION TITLE
iOS Deep Links With Chrome As Default Browser

### DISCUSSION THREAD

**Discussion by bluetoken-luke** [ionic-team/capacitor#4368] [Q&A]
We are running into an issue with our capacitor app and deep linking on iOS.  I realize this is very likely an iOS issue and not a capacitor issue but we've been unable to locate a solution and I'm hoping others may have run into it and can share how they solved it.

When we created our app, we added deep linking as documented here https://capacitorjs.com/docs/v3/guides/deep-links.  Everything seemed to work great and as expected.  Until we ran into a user who was using a browser other than Safari as their default in iOS (in our case it was Chrome, which I assume is pretty common and thus this is a fairly large issue for us).  If the user has a browser other than Safari as their default, deep linking does not work.  If they use Safari, everything works as expected.

We've been looking for solutions to this and have tried a few suggestions such as the second answer here https://stackoverflow.com/questions/64572364/ios-14-universal-links-broken-on-default-browser-other-than-safari but nothing has worked. 

Has anyone else run into this and if so has anyone else solved this issue?

**Comment by sniii** [comment]
Hi Luke

I can acknowledge that my team / project is in **exactly** the same boat as you, unfortunately.
We actually use the universal links / deep linking as part of our login flow, so it's very critical that it just works.

Fortunately for us, it's an internal app in a big company, where everyone is using iPhones and most likely 99/100 users will have Safari installed as default. But we do have one or two cases of people who couldn't use the app because they have Chrome as their default iOS system browser.

I've just been tasked to look into a possible fix for this, but I fear I'll run into the same conclusion as you, i.e. no workable solution.

So, this is a shoutout to the Ionic team; what can we do if we depend on universal links and user has Chrome set as default iOS system browser? Is there a proper way to detect that a default system browser other than Safari is enabled, at least then we can prompt the user?

**Comment by vojto** [comment]
I've come across this issue, and I've been able to isolate it in Chrome like this:

* Create html file on your domain (domain is linked with the iOS app - so clicking on it should open the app)
* Add a link to this html file: Href of this link should go back to the domain

If you open such page in Chrome, and click on the link, nothing will happen. (It won't open the ios app)

* Now try creating a new html file hosted somewhere else. 
* Add a link pointed to your domain

If you click on this link now, it will open the app as it should

---

My solution was to use Firebase Dynamic links that you can easily construct:

<img width="722" alt="CleanShot 2021-09-28 at 11 45 58@2x" src="https://user-images.githubusercontent.com/50943/135064156-dd457e3c-0e2d-4adc-9ddb-7c1acdbd4b18.png">

Now you will be pointed to another domain, so clicking on a link like that will open the app. (The dynamic link domains must also be associated with your app.)

**Comment by seejamescode** [comment]
Hey there! Did either of y'all find a fix to support iOS Chrome for deep links without using a separate domain?

I [found a solution](https://developer.apple.com/forums/thread/665025) to add an `LSApplicationQueriesSchemes` array with a value of `https`, but that didn't work for me.

**Reply by demiluke** [reply]
Unfortunately we did not.  We eventually moved on from capacitor, not specifically due to this.  Strategy changed a bit and native makes a bit more sense for us where we are now/going.

**Reply by joelstevick** [reply]
@demiluke would you be kind enough to share (what you are comfortable sharing) about why you switched from Ionic/capacitor to native?  

We are in a similar decision-point.  We are considering an approach that involves handling the deep-link logic as native code and then passing the route to our Ionic logic via a custom plugin.

Does this make sense to you?

Thanks!

**Comment by Bejasc** [comment]
I'm working with Xamarin.Forms, but noticing the same issue as what's described here (having a little trouble finding much else). 

Same situation, a user with Chrome set as default browser is _not_ handling universal links. 

Has anyone in the ionic/capacitor community also experienced this, and found any solutions - that may provide a hint for where to look on my end?

**Reply by maciejwitalinski** [reply]
same issue here. Tried to integrate keycloak login and using links to go back to app. Doesnt work with chrome as the default browser.

**Reply by NagarajuKamatham** [reply]
@maciejwitalinski we are also in same situation, Have you found any solution?

**Reply by Yonkokilasi** [reply]
Did you find a solution. it's been more than a few years. I am now experiencing this issue

**Reply by OliOnline** [reply]
Yes look down here in this threat, or straight here https://github.com/ionic-team/capacitor/discussions/4368#discussioncomment-11415553

**Comment by inouiw** [comment]
I can confirm universal links do only work in Safari, not in Chrome

The documentation also only mentions Safari:

> With universal links, users open your app when they click links to your website within Safari and [WKWebView](https://developer.apple.com/documentation/webkit/wkwebview?language=objc), and when they click links that result in a call to:
> 
> [open(_:options:completionHandler:)](https://developer.apple.com/documentation/uikit/uiapplication/1648685-open?language=objc) in iOS and tvOS
> 
> [openSystemURL(_:)](https://developer.apple.com/documentation/watchkit/wkextension/1628224-opensystemurl?language=objc) in watchOS
> 
> [open(_:withApplicationAt:configuration:completionHandler:)](https://developer.apple.com/documentation/appkit/nsworkspace/3172702-open?language=objc) in macOS
> 
> [openURL](https://developer.apple.com/documentation/SwiftUI/EnvironmentValues/openURL?language=objc) in SwiftUI

see https://developer.apple.com/documentation/xcode/allowing-apps-and-websites-to-link-to-your-content

**Comment by fikretkurtiiev** [comment]
2024 is still the same issue...
@vojto is your solution relevant nowadays?

**Reply by OliOnline** [reply]
There is a solution, but I can't find it.. AirBnb got it working, they seem to have custom redirections so the link can open the app...

**Reply by fikretkurtiiev** [reply]
@OliOnline we managed this by adding an extra parking page with a button, that the user must click, to make things work. But even after that, in some browsers, the user still must select that he wants to open always in the app, not the web.
One notice here, we added that parking page on a different subdomain (this is very important). It's will be like redirection between different domains (subdomains), otherwise, it will not work.

**Reply by kiran99898** [reply]
with this approach, can we pass data from browser to app? I was to able to open app only.

**Reply by fikretkurtiiev** [reply]
@kiran99898 we are using URL params and this covers our needs.

**Reply by dsrodrigues** [reply]
@OliOnline, were you successful in finding the AirBnb solution? I'm having the same problem.

**Reply by OliOnline** [reply]
@dsrodrigues yes got it working by setting up a subdomain ( something like "app.your-website.com" ) and register the subdomain has a deeplink in Xcode. 
On the website, you can share a url like www.your-website.com/my-deep-link and on this page, there is a button with href "app.your-website.com/my-deep-link". It works on IOS, use click on the share URL, it open a webpage on their default browser and from there they click the button to open your app.


On Android, you need to set a custom schema in android studio ( something like "your-website" ) which will result in href on the button to be "your-website://my-deep-link"

**Comment by JonahSayyed** [comment]
https://app.urlgeni.us/deeplink-tester

check the website guys.if you opened this in chrome browser and paste out deep link url in the field and press on click here.it opens up the app.How are they handeling can anybody tell me.

**Comment by RobinKumar5986** [comment]
Hi I am also facing the same problem dose any one found any solution ??

**Comment by AMER106** [comment]
I see everyone is facing the same problem, but can anyone please provide the solution for it. and i also see, when user clicks on the use my invite code button, the page is just refreshing instead of taking the user to the app.

**Comment by kyllerss** [comment]
I was running into this issue when implementing an OAuth2 sign-in flow. After going through the Apple App Store review process, they required me to use their embedded browser for this process. This became a non-issue at that point. If you are doing OAuth2, bear in mind this requirement might be coming your way.
